### PR TITLE
feat(chrome-extension): Add support for background jobs

### PR DESCRIPTION
**Pull request by tmilewski**
## Description

Adds (early) support for Browser Extension background jobs/service workers.

Example:

```ts
import { __unstable__createClerkClient } from '@clerk/chrome-extension/background';

const publishableKey = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY || '';

async function getToken() {
  const clerk = await __unstable__createClerkClient({ publishableKey /*, syncSessionWithTab: true */ });
  return await clerk.session?.getToken();
}

// NOTE: A runtime listener cannot be async.
//       It must return true, in order to keep the connection open and send a response later.
chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
  getToken().then(token => sendResponse({ token }));
  return true;
});
```

Fixes ECO-117

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

**Comment by tmilewski** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @tmilewski - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.0.5-snapshot.v8fc1f9a |
| @clerk/backend | 1.5.0-snapshot.v8fc1f9a |
| @clerk/chrome-extension | 1.2.0-snapshot.v8fc1f9a |
| @clerk/clerk-js | 5.11.0-snapshot.v8fc1f9a |
| @clerk/elements | 0.11.0-snapshot.v8fc1f9a |
| @clerk/clerk-expo | 2.0.1-snapshot.v8fc1f9a |
| @clerk/express | 0.0.22-snapshot.v8fc1f9a |
| @clerk/fastify | 1.0.24-snapshot.v8fc1f9a |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/localizations | 2.5.3-snapshot.v8fc1f9a |
| @clerk/nextjs | 5.2.9-snapshot.v8fc1f9a |
| @clerk/clerk-react | 5.3.0-snapshot.v8fc1f9a |
| @clerk/remix | 4.2.8-snapshot.v8fc1f9a |
| @clerk/clerk-sdk-node | 5.0.21-snapshot.v8fc1f9a |
| @clerk/shared | 2.4.2-snapshot.v8fc1f9a |
| @clerk/tanstack-start | 0.1.12-snapshot.v8fc1f9a |
| @clerk/testing | 1.2.4-snapshot.v8fc1f9a |
| @clerk/themes | 2.1.15-snapshot.v8fc1f9a |
| @clerk/types | 4.10.0-snapshot.v8fc1f9a |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.0.5-snapshot.v8fc1f9a --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.5.0-snapshot.v8fc1f9a --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.2.0-snapshot.v8fc1f9a --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.11.0-snapshot.v8fc1f9a --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.11.0-snapshot.v8fc1f9a --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.0.1-snapshot.v8fc1f9a --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.22-snapshot.v8fc1f9a --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.24-snapshot.v8fc1f9a --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@2.5.3-snapshot.v8fc1f9a --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.2.9-snapshot.v8fc1f9a --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.3.0-snapshot.v8fc1f9a --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.8-snapshot.v8fc1f9a --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.21-snapshot.v8fc1f9a --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.4.2-snapshot.v8fc1f9a --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.1.12-snapshot.v8fc1f9a --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.2.4-snapshot.v8fc1f9a --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.15-snapshot.v8fc1f9a --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.10.0-snapshot.v8fc1f9a --save-exact
```

**Comment by tmilewski** [comment]
!snapshot

**Comment by tmilewski** [comment]
!snapshot

**Comment by tmilewski** [comment]
!snapshot

**Comment by tmilewski** [comment]
!snapshot snapshot

**Comment by tmilewski** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @tmilewski - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.0.6-snapshot.v0a71e8f |
| @clerk/backend | 1.5.1-snapshot.v0a71e8f |
| @clerk/chrome-extension | 1.2.0-snapshot.v0a71e8f |
| @clerk/dev-cli | 0.0.5-snapshot.v0a71e8f |
| @clerk/elements | 0.12.0-snapshot.v0a71e8f |
| @clerk/express | 0.0.23-snapshot.v0a71e8f |
| @clerk/fastify | 1.0.25-snapshot.v0a71e8f |
| gatsby-plugin-clerk | 5.0.0-snapshot.v0a71e8f |
| @clerk/nextjs | 5.2.11-snapshot.v0a71e8f |
| @clerk/remix | 4.2.9-snapshot.v0a71e8f |
| @clerk/clerk-sdk-node | 5.0.22-snapshot.v0a71e8f |
| @clerk/tanstack-start | 0.1.13-snapshot.v0a71e8f |
| @clerk/testing | 1.2.5-snapshot.v0a71e8f |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.0.6-snapshot.v0a71e8f --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.5.1-snapshot.v0a71e8f --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.2.0-snapshot.v0a71e8f --save-exact
```
`@clerk/dev-cli`
```sh
npm i @clerk/dev-cli@0.0.5-snapshot.v0a71e8f --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.12.0-snapshot.v0a71e8f --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.23-snapshot.v0a71e8f --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.25-snapshot.v0a71e8f --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-snapshot.v0a71e8f --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.2.11-snapshot.v0a71e8f --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.9-snapshot.v0a71e8f --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.22-snapshot.v0a71e8f --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.1.13-snapshot.v0a71e8f --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.2.5-snapshot.v0a71e8f --save-exact
```

File: .changeset/silver-points-cheer.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+---
+'@clerk/chrome-extension': minor
+---
```
Comment (octoper) [inline .changeset/silver-points-cheer.md:3]: ❓ Is there any specific reason having two changesets?

Comment (tmilewski) [reply .changeset/silver-points-cheer.md:3]: @octoper Yeah, it was back when changeset releases weren't working. I can remove the one.

**Review by octoper — APPROVED** [review]
💯

**Comment by flazouh** [comment]
Hello everyone,

I hope you're all doing well. Could someone kindly assist me with a little challenge I'm facing? I'm currently working on storing my Clerk session within my background script. I have a button on my extension page that triggers the following code:

```tsx
<button
  onClick={() => {
    clerk.openSignIn();
  }}
>
  <div className="flex items-center gap-2">
    <GoogleIcon />
    <span>Google</span>
  </div>
</button>
```

When clicked, it opens a modal, which isn't ideal since I'd prefer to directly initiate the Google authentication flow. Nevertheless, it currently opens this modal:

Screenshot 2024-08-26 at 17 28 26

After clicking the Google button and completing the authentication flow, it redirects me to this page:

Screenshot 2024-08-26 at 17 29 04

Ideally, I would like to return to my Chrome extension page and have a callback to store the received token in local storage, so it can be shared across my extension.

I've been through numerous Discord posts and Clerk documentation but have yet to find a solution. Here is the type definition I've been working with:

```ts
type SignInProps = RoutingOptions & {
    forceRedirectUrl?: string | null;
    fallbackRedirectUrl?: string | null;
    signUpUrl?: string;
    appearance?: SignInTheme;
    initialValues?: SignInInitialValues;
} & TransferableOption & SignUpForceRedirectUrl & SignUpFallbackRedirectUrl & LegacyRedirectProps & AfterSignOutUrl;
```

Setting these URLs, for example, results in a "chrome://extension/invalid" URL when set to `window.location.href`.

Any guidance or suggestions you could offer would be greatly appreciated. Thank you in advance for your help!

Best regards

**Pull request merged**